File: 2939-Maximum Xor Product.py
class Solution:
    def maximumXorProduct(self, a: int, b: int, n: int) -> int:
        kMod = 10**9+7
        for i in range(n-1, -1, -1):
            bit = 1 << i
            if a & bit == 0 and b & bit == 0:
                a^=bit
                b^=bit
            elif a & bit == 0 or b & bit == 0:
                a &= ~bit    # clear the ith bit of a
                b &= ~bit    # clear the ith bit of b
                if a >= b:   # choose to put the ith bit in a or in b. If a is larger, put in b
                    b |= bit
                else:
                    a |= bit
        
        return (a%kMod)*(b%kMod)%kMod

File: C++/2939-Maximum Xor Product.cpp
class Solution {
public:
    int maximumXorProduct(long long a, long long b, int n) {
        int kMod = 1e9+7;
        for (int i = n-1; i>=0; i--){
            bool isOneA = (1L<<i)&a;  // long is typically at least 64 bits in size, so it can handle i up to 63 without issue.
            bool isOneB = (1L<<i)&b;
            if (isOneA && isOneB) continue;
            else if (!isOneA && !isOneB) {
                a |= 1L<<i;
                b |= 1L<<i;
            }
            else {
                long long tmpA = a & (~(1L<<i)); // make the ith from 1 to 0
                long long tmpB = b & (~(1L<<i));
                if (tmpA >= tmpB) {
                    b = tmpB | (1L<<i);
                    a = tmpA;
                } else {
                    a = tmpA | (1L<<i);
                    b = tmpB;
                }
            }
        }
        return (a%kMod)*(b%kMod)%kMod;
    }
};